streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -451,15 +451,23 @@ public void restore() {
     }
 
     private void maybeUpdateLimitOffsetsForStandbyChangelogs() {
-        // for standby changelogs, if the interval has elapsed and there are buffered records not applicable,
-        // we can try to update the limit offset next time.
-        if (updateOffsetIntervalMs < time.milliseconds() - lastUpdateOffsetTime) {
-            final Set<ChangelogMetadata> standbyChangelogs = changelogs.values().stream()
-                .filter(metadata -> metadata.stateManager.taskType() == Task.TaskType.STANDBY)
-                .collect(Collectors.toSet());
-            for (final ChangelogMetadata metadata : standbyChangelogs) {
-                if (!metadata.bufferedRecords().isEmpty()) {
-                    updateLimitOffsets();
+        // we only consider updating the limit offset for standbys if we are not restoring active tasks
+        if (state == ChangelogReaderState.STANDBY_UPDATING &&
+            updateOffsetIntervalMs < time.milliseconds() - lastUpdateOffsetTime) {
+
+            // for standby changelogs, if the interval has elapsed and there are buffered records not applicable,
+            // we can try to update the limit offset as either committed offset for source changelog partitions;
```

Comment:
```suggestion
            // when the interval has elapsed we should try to update the limit offset for standbys reading from
            // a source changelog with the new committed offset, unless there are no buffered records since 
            // we only need the limit when processing new records
```

Reply:
Q: if we skip updating the limit offsets because the buffer is empty, and then fetch new records immediately after, we'll still have to wait another commit interval to update the limit offsets. This means the standbys are potentially  lagging behind by up to the commit interval. Do we think that's worth the optimization of skipping the limit offset update?. Note I've seen a number of users report setting the commit interval quite large for various reasons (eg avoid flushing the memtables prematurely).

Maybe we can find some middle ground by updating the limits before the commit interval if we do get some buffered records and skipped updating previously?

Reply:
We only update `lastUpdateOffsetTime = time.milliseconds();` when we've indeed refreshed our committed offsets, so if we did not refresh the committed offsets the timer would not be reset. I think that should be sufficient for covering your concern?

Reply:
Cool, yeah that addresses my concern 👍